### ISSUE TITLE
[P1][Apple Silicon Runtime] Unified Memory / Thermal / Power / Sleep Guardrails / MLX Process Recovery

### ISSUE THREAD

**Issue by dasomel** [dasomel/kubemetal#13]
## 목표
KubeMetal의 Apple Silicon 전용 runtime에서 **unified memory·thermal·power·sleep·process failure**를 MLX fine-tuning/serving lifecycle과 연결해 장시간 로컬 작업에서도 안전하게 수렴하도록 한다.

현재 KubeMetal은 K8s control plane을 Colima VM에서 실행하고 실제 GPU-bound MLX 작업은 macOS host process에서 실행한다. Apple Silicon Metal GPU는 Linux VM에 passthrough할 수 없기 때문에 host runtime 제약이 핵심 운영 경계다. 

## 설계 원칙
- Apple unified memory를 NVIDIA VRAM과 동일하게 해석하지 않는다.
- memory pressure와 thermal state를 별도 safety signal로 관리한다.
- thermal serious/critical, memory pressure, battery/power, sleep/lock 상황에서 task별 정책을 적용한다.
- fine-tuning과 serving은 서로 다른 safety policy를 가질 수 있다.
- host MLX process가 비정상 종료되어도 control-plane state와 MLflow run state가 일치해야 한다.
- 자동 중단/재개보다 checkpoint/resume을 우선한다.
- 무리한 batch downscaling 등은 model/task compatibility에 따라 명시적 정책으로만 수행한다.

## Scope
- unified memory usage/pressure monitoring
- GPU-related memory vs system memory semantics
- thermal state monitoring and threshold policy
- power/battery/AC state awareness
- sleep/lock/screen state interaction policy
- training pause/resume policy
- inference serving degradation policy
- checkpoint cadence and restart recovery
- MLX process heartbeat / liveness / exit reason
- orphan process detection/cleanup
- MLflow run state reconciliation after process failure
- host process ↔ K8s control-plane lifecycle correlation
- recovery after macOS reboot / logout / sleep-wake
- resource reservation and admission for concurrent MLX jobs
- metrics: memory pressure, thermal state, pause time, restart count, recovery time
- offline/reproducible failure-injection tests

## Acceptance Criteria
- [ ] unified memory usage와 memory pressure를 별도 metric/state로 표시한다
- [ ] Apple Metal telemetry capability metadata가 #1과 일관되게 적용된다
- [ ] thermal serious/critical에서 task-specific safety action이 deterministic하게 동작한다
- [ ] battery/AC 상태와 장시간 fine-tuning 정책을 연결할 수 있다
- [ ] sleep/wake 후 running job의 상태를 감지하고 안전한 resume/recovery 경로를 제공한다
- [ ] MLX process crash/kill 이후 duplicate training/serving process가 남지 않는다
- [ ] MLflow run state가 실제 host process 결과와 수렴한다
- [ ] checkpoint와 last-known-good state를 사용해 training resume이 가능하다
- [ ] concurrent local jobs에 대해 memory/thermal admission guardrail을 적용할 수 있다
- [ ] reboot/restart 후 pipeline state와 artifact state를 복구할 수 있다
- [ ] 모든 pause/restart/recovery event가 correlation ID와 evidence를 가진다
- [ ] failure-injection과 long-run soak test가 CI/validation profile에 포함된다

## 검증
1. 64GB Apple Silicon에서 장시간 MLX fine-tuning을 수행해 memory/thermal pressure를 측정한다.
2. thermal serious/critical 상태를 주입해 pause/resume 정책을 검증한다.
3. memory pressure를 유도해 unsafe OOM 대신 configured safety action이 실행되는지 확인한다.
4. battery/AC 상태를 변경해 long-running job policy를 검증한다.
5. training 중 process kill을 주입하고 checkpoint 기반 resume과 MLflow 상태 수렴을 검증한다.
6. serving process crash 후 자동 복구와 duplicate process 방지를 검증한다.
7. macOS sleep/wake를 수행해 job/process/control-plane state를 비교한다.
8. macOS reboot 후 pipeline/artifact/run state recovery를 검증한다.
9. 두 개 이상의 MLX workload를 동시에 실행해 admission/guardrail을 검증한다.
10. 장시간 soak test에서 memory leak, thermal runaway, orphan process가 없는지 확인한다.

## Related
- KubeMetal #1 GPU Telemetry Interoperability — Apple Metal / Remote NVIDIA
- KubeMetal #12 Apple Silicon Model Lifecycle / Evaluation Gate / MLX Artifact Registry / Deployment Rollback

## Narwhal 연계
- Narwhal #105 Storage SLO / Capacity Planning
- Narwhal #112 Error Budget / SLO Governance
- Narwhal #134 AI/ML Governance reference

## 범위 제외
- NVIDIA GPU runtime 자체 구현
- Metal kernel/MLX 자체 개발
- macOS power management 자체 구현
- 무조건적인 autonomous resource termination

**Comment by dasomel** [comment]
## 결함 3건 수정 — `35c6be3`

이 이슈 범위(가드레일 / MLX 프로세스 복구)에서 **실제 오동작 3건**을 찾아 고쳤습니다. 셋 다 같은 형태입니다 — 앱이 OS가 동의하지 않는 상태를 보고합니다(D22).

### 1. 학습 중지 버튼이 "오류"를 띄웠습니다

사용자가 중지를 누르면 `Training process exited abnormally` 오류가 표시됐습니다.

`kill_mlx_process`가 시그널을 먼저 보내고 `killed`를 나중에 썼는데, **`terminate_pid`는 SIGTERM과 SIGKILL 사이에 1초를 잡니다**(`mlx.rs:618`). 그 사이 `run_training_reader`가 종료를 관측해 `finalize_training`을 호출하고 `error`를 먼저 씁니다. 그다음 kill이 도달해도 `!= "error"` 가드에 걸려 `killed` 갱신이 스킵됩니다.

드문 경합이 아닙니다 — 그 1초 때문에 **거의 항상** 발생합니다.

→ 시그널 **전에** 의도를 기록합니다. `finalize_training`의 기존 가드가 그대로 보호막이 됩니다. 새 가드가 필요했던 게 아니라 쓰는 시점이 늦었던 것입니다.

### 2. 일시정지 중 밖에서 죽으면 영원히 "일시정지 중"

`paused_*` 상태에서 OOM killer 등으로 프로세스가 죽어도 화면이 계속 "일시정지 중"이었습니다. `finalize_training`이 `status == "running"`일 때만 동작했기 때문입니다.

→ 판정을 **"종착 상태가 아니면 기록한다"**(`should_record_exit`)로 바꿨습니다. 1번과 상호의존입니다 — 한쪽만 고치면 다른 쪽이 깨집니다.

### 3. `signal_pid`가 kill 실패를 삼켰습니다

```rust
tokio::task::spawn_blocking(move || unsafe { libc::kill(-(pid as i32), sig); })
    .await
    .map_err(|e| format!("failed to send signal: {e}"))
```

`map_err`는 `spawn_blocking`의 **JoinError만** 잡습니다. `libc::kill`의 반환값은 버려집니다. 죽은 pid에 SIGSTOP을 보내고도 `Ok(())`가 돌아와, 호출부가 존재하지 않는 프로세스를 `paused_*`로 표시했습니다.

→ 반환값을 검사하고, **pid 0을 거부**합니다. `kill(0, …)`은 앱 자신의 프로세스 그룹으로 가므로 SIGSTOP이면 앱이 스스로 얼어붙습니다.

### 배경 원인

`TrainingStatus.status` 주석이 실제 8개 값 중 4개만 적고 `paused*`를 전부 빠뜨리고 있었습니다 — `"running"`이 비종료 집합 전체처럼 보이게 만든 배경입니다. 함께 고쳤습니다.

### 테스트

`should_record_exit`의 양방향(`paused*`는 기록 / `done`·`error`·`killed`는 보호), `signal_pid`의 pid 0 거부와 죽은 프로세스 실패 보고(실제 자식을 띄워 **살아 있을 때 성공·회수 후 실패**를 모두 확인). `make verify` 77 통과.

### 남은 것 (이슈 유지)

MLX 프로세스 **복구/재조정**은 미착수입니다 — 지금 한 것은 상태를 정직하게 보고하는 데까지입니다. 죽은 학습의 자동 재시작, 체크포인트에서의 재개는 별개 작업입니다.

**Comment by dasomel** [comment]
D28(발열 `serious` 임계, opt-in)·D16(메모리 압력 가드레일)이 이미 이 이슈와 겹치는 부분(발열 정책)을 커버합니다. 나머지 대부분은 미구현이고 규모가 큽니다: MLX 프로세스 heartbeat/orphan 감지, 크래시 후 MLflow run state 수렴, checkpoint/resume, sleep/wake/reboot 복구, battery/AC 연동 정책, concurrent job admission(#32와 연관), failure-injection/soak 테스트.

이번 세션 범위 밖입니다. 여러 하위 작업으로 쪼개 오너가 우선순위를 정해야 합니다. 코드 변경 없이 코멘트만 남깁니다.

**Comment by dasomel** [comment]
**권장: 유지(open), 후순위, 쪼개기 권장.** heartbeat/orphan 감지·resume·sleep 복구 등 하위 작업이 많아 하나의 이슈로 두기보단 쪼개는 것을 권장합니다.

**Comment by dasomel** [comment]
**"MLX Process Recovery" 첫 슬라이스 구현 완료** — 고아 프로세스 탐지.

실측 gap: 앱이 크래시/강제종료되면 `MlxState`(인메모리)는 새 실행에서 빈 상태로 초기화되지만, 아직 살아있는 mlx_lm/mlx_vlm 자식 프로세스는 아무도 감시하지 않는 고아가 됩니다 — 사용자는 "학습 없음"을 보지만 실제로는 GPU/메모리를 계속 먹는 프로세스가 남습니다.

- `src-tauri/src/commands/mlx.rs`: 학습/서빙 스폰 직후 `~/.kubemetal/mlx-{training,serving}.pid`에 pid 기록, `child.wait()` 완료 시 marker 제거. 새 IPC `check_for_orphaned_mlx_processes()`가 marker의 pid 생존 여부(`kill -0`)를 확인 — 죽어있으면 지어내지 않고 marker만 정리(D22), 살아있으면 보고만 하고 자동으로 죽이지 않습니다(사용자가 학습 중인 걸 모르고 재시작했을 수 있음).
- **알려진 한계 명시**: PID 재사용 시 오탐 가능 — 완전 해소는 `ps` 커맨드라인 대조나 별도 슈퍼바이저가 필요해 스코프 밖으로 남겼습니다.
- 테스트 5개, `cargo test`(130 passed)/`clippy`/IPC 게이트 통과. 커밋: `beb4a70`.

**남은 gap(이슈 유지)**: sleep/wake/reboot 복구, battery 연동 정책, 프런트 UI 통합(탐지된 고아를 사용자에게 보여주고 kill 옵션 제공), concurrent job admission(#31/#32에서 이미 다룸), failure-injection/soak 테스트.

**Comment by dasomel** [comment]
**"크래시 후 MLflow run state 수렴" 슬라이스 추가 구현 완료.**

`finetune_wrapper.py`는 정상 흐름(성공/실패 모두)에서 자기 `reporter.end_run(...)`을 먼저 부른 뒤에만 종료합니다 — 그런데 `kill_mlx_process`가 보내는 SIGTERM/SIGKILL로 죽으면 그 코드에 도달하지 못해 MLflow run이 `RUNNING`에 영원히 머뭅니다(사용자가 학습을 멈췄는데 MLflow UI엔 계속 돌아가는 것처럼 보임).

- `finetune_wrapper.py`: `start_run` 성공 직후 실제 run_id를 stdout 이벤트로 보고.
- `mlx.rs`: `TrainingStatus.mlflow_run_id`로 수신. `mlflow_reconciliation_decision`(순수 함수) — wrapper가 이미 done/error를 보고했으면 손대지 않고, **시그널로 죽었고 아직 보고 전이면**(status·run_id 무관하게) MLflow `runs/update`를 `KILLED`로 호출(`external_command("curl")` — modelhub.rs의 기존 관례 재사용, 새 HTTP 클라이언트 없음). 실패는 학습 종료 처리를 막지 않는 best-effort.
- 테스트 5개, `cargo test`(130+ passed)/`clippy`/IPC 게이트 통과. 커밋: `bbc4a17`.

**투명하게 공유합니다**: 이 부분은 처음에 Codex(`omc ask codex`)에게 위임했는데, Codex가 계약 테스트 5개까지만 쓰고 실제 구현 전에 OpenAI 사용량 한도(자기 안에서 이 저장소의 `agyp` 위임 체계를 재귀적으로 호출하려다 얽힌 것도 한 원인)로 중단됐습니다. 컴파일 안 되는 상태였던 걸 검토 중 잡아서, 이미 잘 설계된 테스트 계약을 이어받아 제가 직접 구현했습니다.

**여전히 남은 gap**: sleep/wake/reboot 복구, battery 연동 정책, 프런트 UI 통합, failure-injection/soak 테스트.

**Comment by dasomel** [comment]
고아 MLX 프로세스 탐지(`beb4a70`)와 MLflow run KILLED 수렴(`bbc4a17`)은 이전 스레드에서 '구현됨'으로 언급됐지만, **main에 들어간 적이 없습니다**(유실 커밋, `docs/mistakes-log.md` 2026-09-23 행 참고). 현재 `mlx.rs`와 충돌이 커서 PR #99에서는 제외했습니다. 재반영은 다음 단계입니다.

**Comment by dasomel** [comment]
PR #100에서 유실 커밋 beb4a70(고아 MLX 프로세스 탐지)과 bbc4a17(MLflow KILLED 수렴)을 현재 main에 재이식했습니다. Codex 리뷰의 HIGH 1건과 MED 7건도 함께 반영했습니다. 주요 내용: run_id를 같은 pid에만 귀속, 프로세스별 marker를 temp+rename으로 기록, 명령줄 토큰 판정, curl timeout과 HTTP 2xx 판정, 새 IPC `check_for_orphaned_mlx_processes`는 `Result<OrphanScan>` 반환(자동 kill 없음). 단위 테스트와 `make verify`는 통과했습니다. 실제 앱 크래시 후 탐지와 실제 MLflow 수렴은 아직 검증하지 않았습니다. **여전히 남은 gap**: sleep/wake/reboot 복구, battery 정책, 프런트 UI 통합(현재 IPC를 호출하는 곳 없음), failure-injection/soak 테스트. 그래서 PR이 머지돼도 이 이슈는 열어 둡니다.

**Comment by dasomel** [comment]
PR #103이 머지됐습니다(GPT 팀 레인. Codex가 한도로 멈춘 뒤 native 레인이 완성했고, opus 승인을 받았습니다).
- **#13**: 고아 스캔이 현재 세션이 추적 중인 pid를 제외합니다(marker는 보존). 남은 것: sleep/wake/reboot 복구, battery 정책, 프런트 UI 통합, failure-injection/soak 테스트.
